v2/internal/identity/credential_provider.go

```diff
@@ -0,0 +1,260 @@
+
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/reconcilers"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/core"
+)
+
+const (
+	// #nosec
+	globalCredentialSecretName = "aso-controller-settings"
+	// #nosec
+	NamespacedSecretName = "aso-credential"
+)
+
+// Credential describes a credential used to connect to Azure
+type Credential struct {
+	TokenCredential azcore.TokenCredential
+	CredentialFrom  types.NamespacedName
+	SubscriptionID  string
+
+	// secretData contains the secret
+	secretData map[string][]byte
+}
+
+func (c *Credential) SecretsEqual(other *Credential) bool {
+	return reflect.DeepEqual(c.secretData, other.secretData)
+}
+
+func NewDefaultCredential(tokenCred azcore.TokenCredential, namespace string, subscriptionID string) *Credential {
+	return &Credential{
+		TokenCredential: tokenCred,
+		SubscriptionID:  subscriptionID,
+		CredentialFrom:  types.NamespacedName{Namespace: namespace, Name: globalCredentialSecretName},
+	}
+}
+
+type CredentialProvider interface {
+	GetCredential(ctx context.Context, obj genruntime.MetaObject) (*Credential, error)
+}
+
+type credentialProvider struct {
+	globalCredential *Credential
+	kubeClient       kubeclient.Client
+}
+
+func NewCredentialProvider(
+	globalCredential *Credential,
+	kubeClient kubeclient.Client,
+) CredentialProvider {
+	return &credentialProvider{
+		kubeClient:       kubeClient,
+		globalCredential: globalCredential,
+	}
+}
+
+// GetCredential returns the credential used to manage the obj. It performs credential lookup according to the following
+// hierarchy:
+// 1. Per-Resource credential specified as an annotation (serviceoperator.azure.com/credential-from) on the obj
+// 2. Per-Namespace credential provided at the namespace scope ("aso-credential" secret in the obj's namespace)
+// 3. Global credential for the operator.
+//
+// If no matching credential can be found, an error is returned.
+func (c *credentialProvider) GetCredential(ctx context.Context, obj genruntime.MetaObject) (*Credential, error) {
+	cred, err := c.getCredentialFromAnnotation(ctx, obj, reconcilers.PerResourceSecretAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	if cred != nil {
+		return cred, nil
+	}
+
+	// Namespaced secret
+	cred, err = c.getCredentialFromNamespaceSecret(ctx, obj.GetNamespace())
+	if err != nil {
+		return nil, err
+	}
+	if cred != nil {
+		return cred, nil
+	}
+
+	if c.globalCredential == nil {
+		return nil, errors.New("global credential not configured, you must use either namespaced or per-resource credentials")
+	}
+	// If not found, default is the global credential
+	return c.globalCredential, nil
+}
+
+func (c *credentialProvider) getCredentialFromNamespaceSecret(ctx context.Context, namespace string) (*Credential, error) {
```

Comment:
Suggest adding a doc comment explaining what this method does

Reply:
Done